### PR TITLE
Sandbox missing paths alternative

### PR DESCRIPTION
**Pull request by ThetaSinner**
### TODO:
- [ ] CHANGELOGs updated with appropriate info
- [ ] All code changes are reflected in docs, including module-level docs

### CODE REVIEW

File: crates/hc_sandbox/src/cmds.rs [hunk 1]
```diff
@@ -93,54 +93,68 @@ pub enum NetworkType {
 
 #[derive(Debug, Parser, Clone)]
 pub struct Existing {
-    /// Paths to existing sandbox directories.
-    /// For example `hc sandbox run -e=/tmp/kAOXQlilEtJKlTM_W403b,/tmp/kddsajkaasiIII_sJ`.
-    #[arg(short, long, value_delimiter = ',')]
-    pub existing_paths: Vec<PathBuf>,
```
Comment (ThetaSinner) [inline crates/hc_sandbox/src/cmds.rs:99]: Removed, the sandbox will always work with paths it's created in the `.hc` file in the current directory.

File: crates/hc_sandbox/src/cmds.rs [hunk 2]
```diff
@@ -93,54 +93,68 @@ pub enum NetworkType {
...
-
     /// Run all the existing conductor sandboxes specified in `$(pwd)/.hc`.
-    #[arg(short, long, conflicts_with_all = &["last", "indices"])]
+    #[arg(short, long, conflicts_with_all = &["indices"])]
     pub all: bool,
 
-    /// Run the last created conductor sandbox --
-    /// that is, the last line in `$(pwd)/.hc`.
-    #[arg(short, long, conflicts_with_all = &["all", "indices"])]
-    pub last: bool,
```
Comment (ThetaSinner) [inline crates/hc_sandbox/src/cmds.rs:108]: Not commonly used, use all or an index instead

File: crates/hc_sandbox/src/cmds.rs [hunk 3]
```diff
@@ -93,54 +93,68 @@ pub enum NetworkType {
...
         let sandboxes = crate::save::load(std::env::current_dir()?)?;
         if self.all {
-            // Get all the sandboxes
-            self.existing_paths.extend(sandboxes)
-        } else if self.last && sandboxes.last().is_some() {
-            // Get just the last sandbox
-            self.existing_paths
-                .push(sandboxes.last().cloned().expect("Safe due to check above"));
+            // If any of the sandboxes are missing, warn the user
+            if sandboxes.iter().any(Option::is_none) {
```
Comment (ThetaSinner) [inline crates/hc_sandbox/src/cmds.rs:122]: Warn on any missing sandbox directories

File: crates/hc_sandbox/src/cmds.rs [hunk 4]
```diff
@@ -93,54 +93,68 @@ pub enum NetworkType {
...
+            }
+
+            // Return all the valid sandboxes
+            Ok(sandboxes.into_iter().filter_map(|p| p).collect())
         } else if !self.indices.is_empty() {
-            // Get the indices
-            let e = self
+            // If the user specified indices, then all sandboxes must be present. Otherwise, the
+            // user might be referring to a different sandbox than they expect.
+            if sandboxes.iter().any(Option::is_none) {
```
Comment (ThetaSinner) [inline crates/hc_sandbox/src/cmds.rs:132]: Error on any missing sandbox directories

File: crates/hc_sandbox/src/cmds.rs [hunk 5]
```diff
@@ -161,12 +180,12 @@ Before running or calling you need to generate a sandbox.
 You can use `hc sandbox generate` to do this.
 Run `hc sandbox generate --help` for more options."
             );
+            Err(std::io::Error::other("No sandboxes found."))
         }
-        Ok(self.existing_paths)
```
Comment (ThetaSinner) [inline crates/hc_sandbox/src/cmds.rs:165]: Restructured so that each branch returns a value

File: crates/hc_sandbox/src/cmds.rs [hunk 6]
```diff
@@ -248,3 +267,412 @@ impl Default for Create {
 fn try_parse_url2(arg: &str) -> url2::Url2Result<Url2> {
     Url2::try_parse(arg)
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (ThetaSinner) [inline crates/hc_sandbox/src/cmds.rs:272]: AI generated tests, as an example of the balance between code change/testing that I'd prefer to see where code is missing unit tests.

File: crates/hc_sandbox/src/save.rs [hunk 7]
```diff
@@ -78,7 +81,7 @@ pub fn clean(mut hc_dir: PathBuf, sandboxes: Vec<usize>) -> anyhow::Result<()> {
 }
 
 /// Load sandbox paths from the `.hc` file.
-pub fn load(mut hc_dir: PathBuf) -> anyhow::Result<Vec<PathBuf>> {
+pub fn load(mut hc_dir: PathBuf) -> std::io::Result<Vec<Option<PathBuf>>> {
```
Comment (ThetaSinner) [inline crates/hc_sandbox/src/save.rs:84]: Prefer using std::io::Result, since IO errors is all this code can return anyway. Using `anyhow` when consuming the sandbox as a library is difficult. Prefer only using anyhow in binary-only code.

File: crates/hc_sandbox/src/save.rs [hunk 8]
```diff
@@ -87,36 +90,45 @@ pub fn load(mut hc_dir: PathBuf) -> anyhow::Result<Vec<PathBuf>> {
             let path = PathBuf::from(sandbox);
             let config_file_path = ConfigFilePath::from(ConfigRootPath::from(path.clone()));
             if config_file_path.as_ref().exists() {
-                paths.push(path);
+                paths.push(Some(path));
             } else {
                 tracing::error!("Failed to load path {} from existing .hc", path.display());
+                paths.push(None);
```
Comment (ThetaSinner) [inline crates/hc_sandbox/src/save.rs:96]: Keep track of any missing paths for indexing but if the path is invalid, we don't need to keep it.

File: crates/hc_sandbox/src/save.rs [hunk 9]
```diff
@@ -214,3 +226,602 @@ pub async fn release_ports(hc_dir: PathBuf) -> anyhow::Result<()> {
     }
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (ThetaSinner) [inline crates/hc_sandbox/src/save.rs:231]: More generated tests, I did look through these and they look reasonable but would need some more thought about whether they cover edge cases properly if this PR was going anywhere.

**Comment by ThetaSinner** [comment]
Closing, the PR was just an example and not intended to merge. It has served its purpose.
